### PR TITLE
pg_upgrade nonupgradable test detecting views with removed functions

### PR DESCRIPTION
**Pull request by kyeap-vmware**
Views that use removed functions will cause upgrade to fail. This happens during metadata restore on the target cluster because pg_restore will error trying to create a view using functions that do not exist anymore. This is not ideal as we could be many hours into upgrade before a view using a removed function causes pg_upgrade to fail. This check calls a support function on the source cluster to check if views using removed functions exist.

https://github.com/greenplum-db/gpupgrade/pull/902
6x: https://github.com/greenplum-db/gpdb/pull/17152
7x: https://github.com/greenplum-db/gpdb/pull/17153

### CODE REVIEW

File: test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/sql/views_with_removed_functions.sql [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+CREATE VIEW v04 AS SELECT pg_get_partition_def('rank_table'::regclass);
+CREATE VIEW v05 AS SELECT pg_get_partition_def('rank_table'::regclass, true);
+CREATE VIEW v06 AS SELECT pg_get_partition_def('rank_table'::regclass, true, true);
+CREATE VIEW v07 AS SELECT pg_get_partition_rule_def('rank_table'::regclass);
+CREATE VIEW v08 AS SELECT pg_get_partition_rule_def('rank_table'::regclass, true);
+CREATE VIEW v09 AS SELECT pg_get_partition_template_def('rank_table'::regclass, true, true);
+CREATE VIEW v10 AS SELECT pg_catalog.pg_is_xlog_replay_paused();
+CREATE VIEW v11 AS SELECT pg_catalog.pg_last_xlog_receive_location();
+CREATE VIEW v12 AS SELECT pg_catalog.pg_last_xlog_replay_location();
+CREATE VIEW v13 AS SELECT pg_catalog.pg_switch_xlog();
```
Comment (kalensk) [inline test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/sql/views_with_removed_functions.sql:29]: Maybe its just me but it seems like it would be more self documenting if the view was called based on the table. Might also help with debugging and things. Something like `view_pg_current_xlog_insert_location`. Not sure if that is too wordy, but it is more clear than v01

File: test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/sql/views_with_removed_functions.sql [hunk 2]
```diff
@@ -0,0 +1,54 @@
+-- Copyright (c) 2017-2024 VMware, Inc. or its affiliates
+-- SPDX-License-Identifier: Apache-2.0
+
+--------------------------------------------------------------------------------
+-- Create and setup non-upgradeable objects
+--------------------------------------------------------------------------------
+
+CREATE TABLE foo(i int) with (appendonly=true);
```
Comment (kalensk) [inline test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/sql/views_with_removed_functions.sql:8]: Can we rename `foo` to something like` ao_table`?

File: test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/sql/views_with_removed_functions.sql (hunk 1 above)
Comment (kyeap-vmware) [reply test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/sql/views_with_removed_functions.sql:29]: The list of views here containing removed functions are arbitrary and non exhaustive so I chose a naming scheme that was easy to verify at a glance rather than explicit in what the views contained. There are over 100 removed functions. I just chose the subset that was the easiest to test.

File: test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/expected/views_with_removed_functions.out [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+DROP VIEW v03;
+DROP VIEW
+DROP VIEW v02;
+DROP VIEW
+DROP VIEW v01;
+DROP VIEW
+DROP TABLE rank_table;
+DROP TABLE
+DROP TABLE foo;
+DROP TABLE
```
Comment (soumyadeep2007) [inline test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/expected/views_with_removed_functions.out:94]: No need to drop these. The "workaround" section should only contain objects that the customer would need to operate on and nothing more.

**Review by soumyadeep2007 — APPROVED** [review]
Just a minor thing. Create all the objects in a schema to avoid conflicts with other tests (foo/ao_table/v01 etc aren't very unique).

Otherwise LGTM!

File: test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/expected/views_with_removed_functions.out (hunk 3 above)
Comment (kyeap-vmware) [reply test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/expected/views_with_removed_functions.out:94]: > The "workaround" section should only contain objects that the customer would need to operate on and nothing more.

Is there a reason we would need to keep certain objects around after testing? I've been using this section more like a cleanup section. My intention was to keep the tests isolated.

Comment (kyeap-vmware) [reply test/acceptance/pg_upgrade/6-to-7/non_upgradeable_tests/expected/views_with_removed_functions.out:94]: > Create all the objects in a schema to avoid conflicts with other tests

This is also part of the reason why the tests cleanup after themselves. Maybe a "cleanup" section would be more accurate.
